src/symmetry.jl

```diff
@@ -239,3 +239,81 @@ function check_symmetric(basis, ρin; tol=1e-10, symmetries=ρin.basis.model.sym
+This is mainly useful for debug purposes (e.g. in cases we don't want to
+bother thinking about symmetries).
+"""
+function unfold_BZ(basis::PlaneWaveBasis)
+    if all(length.(basis.ksymops_global) .== 1)
+        return basis
+    else
+        kcoords = []
+        for (ik, kpt) in enumerate(basis.kcoords_global)
+            for (S, τ) in basis.ksymops_global[ik]
+                push!(kcoords, normalize_kpoint_coordinate(S * kpt))
+            end
+        end
+        new_basis = PlaneWaveBasis(basis.model,
+                                   basis.Ecut, basis.fft_size, basis.variational,
+                                   kcoords, [[identity_symop()] for _ in 1:length(kcoords)],
+                                   basis.kgrid, basis.kshift, basis.symmetries, basis.comm_kpts)
+    end
+end
+
+# find where in an irreducible basis kpoint `kcoord_unfolded` is handled
+function unfold_mapping(basis_irred, kpt_unfolded)
+    for ik_irred = 1:length(basis_irred.kpoints)
+        kpt_irred = basis_irred.kpoints[ik_irred]
+        for symop in basis_irred.ksymops[ik_irred]
+            if normalize_kpoint_coordinate(symop[1] * kpt_irred.coordinate) ≈
+                normalize_kpoint_coordinate(kpt_unfolded.coordinate) && kpt_unfolded.spin == kpt_irred.spin
+                return ik_irred, symop
+            end
+        end
+    end
+    error("Invalid unfolding of BZ")
+end
+
+function unfold_array(basis_irred, basis_unfolded, data, is_ψ)
+    if basis_irred == basis_unfolded
+        return data
+    end
+    if !(basis_irred.comm_kpts == basis_irred.comm_kpts == MPI.COMM_WORLD)
+        error("Brillouin zone symmetry unfolding not supported with MPI yet")
+    end
+    data_unfolded = similar(data, length(basis_unfolded.kpoints))
+    for ik_unfolded in 1:length(basis_unfolded.kpoints)
+        kpt_unfolded = basis_unfolded.kpoints[ik_unfolded]
+        ik_irred, symop = unfold_mapping(basis_irred, kpt_unfolded)
+        if !is_ψ
+            data_unfolded[ik_unfolded] = data[ik_irred]
+        else
+            @assert normalize_kpoint_coordinate(kpt_unfolded.coordinate) ≈ kpt_unfolded.coordinate
+            data_unfolded[ik_unfolded] = apply_ksymop(symop, basis_irred,
+                                                        basis_irred.kpoints[ik_irred],
+                                                        data[ik_irred])[2]
+        end
+    end
+    data_unfolded
+end
+unfold_array(basis_irred, basis_unfolded, data::Vector{Vector{T}}) where {T <: Number} =
+    unfold_array(basis_irred, basis_unfolded, data, false)
+unfold_ψ(basis_irred, basis_unfolded, ψ) = unfold_array(basis_irred, basis_unfolded, ψ, true)
+
+function unfold_BZ(scfres)
+    basis_unfolded = unfold_BZ(scfres.basis)
+    ψ = unfold_ψ(scfres.basis, basis_unfolded, scfres.ψ)
+    eigenvalues = unfold_array(scfres.basis, basis_unfolded, scfres.eigenvalues)
+    occupation = unfold_array(scfres.basis, basis_unfolded, scfres.occupation)
```

Comment:
I think I would structure this differently. For me these three objects are inherently connected and I think it makes sense to fuse the loop, i.e. to handle them simultaneously. That way you also get rid of this strange `is_\psi` thingy. Also what other k-Point specific objects will we need to transfer in such an unfolding? I don't thing any more.

Reply:
So you'd just merge the unfold_BZ and unfold_array functions? I kind of like the separation. You could have other things, like a partial density, a matrix element for a response computation, etc

Reply:
True, but in this setup you do a lot of duplicate work `unfold_mapping` is computed each `unfold_array` call. I can see us quickly overusing `unfold_BZ`, so I'm not too happy about this structure. If you have the unfolding as the outer loop it's still easy to add an extra line if you have an additional entry in your scfres.

Reply:
Unfold_mapping is cheap. I definitely would wait until we have a better idea of the use cases for this before doing any refactoring. Not convinced merging functions would help here

Reply:
OTOH `is_psi` flag is really annoying and the wrapping it into two function names as well. It feels very hackish and I'm not a fan of this solution. If you swap inner and outer loop than you have that info and do not need that strangle function structure. I think that is the easiest solution and as you say, when we need more we refactor.

Reply:
I've removed the extra functions which were more confusing than anything (trying to pretend this is a high-level API even if it's pretty hacky). But I really don't want to merge the functions : this is already a pain to write and read, having an extra function does help I feel